packages/jest-circus/src/run.js

```diff
@@ -0,0 +1,97 @@
+/**
+ * Copyright (c) 2014-present, Facebook, Inc. All rights reserved.
+ *
+ * This source code is licensed under the BSD-style license found in the
+ * LICENSE file in the root directory of this source tree. An additional grant
+ * of patent rights can be found in the PATENTS file in the same directory.
+ *
+ * @flow
+ */
+
+import type {
+  Test,
+  TestResults,
+  TestContext,
+  Hook,
+  DescribeBlock,
+} from '../types';
+
+const {getState, dispatch} = require('./state');
+const {
+  callAsyncFn,
+  getAllHooksForDescribe,
+  getEachHooksForTest,
+  makeTestResults,
+} = require('./utils');
+
+const run = async (): Promise<TestResults> => {
+  const {topDescribeBlock} = getState();
+  dispatch({name: 'run_start'});
+  await _runTestsForDescribeBlock(topDescribeBlock);
+  dispatch({name: 'run_finish'});
+  return makeTestResults(getState().topDescribeBlock);
+};
+
+const _runTestsForDescribeBlock = async (describeBlock: DescribeBlock) => {
+  dispatch({describeBlock, name: 'run_describe_start'});
+  const {beforeAll, afterAll} = getAllHooksForDescribe(describeBlock);
+
+  for (const hook of beforeAll) {
+    _callHook(hook);
+  }
+  for (const test of describeBlock.tests) {
+    await _runTest(test);
+  }
+  for (const child of describeBlock.children) {
+    await _runTestsForDescribeBlock(child);
+  }
+
+  for (const hook of afterAll) {
+    _callHook(hook);
+  }
+  dispatch({describeBlock, name: 'run_describe_finish'});
+};
+
+const _runTest = async (test: Test): Promise<void> => {
+  const testContext = Object.create(null);
+  const {afterEach, beforeEach} = getEachHooksForTest(test);
+
+  for (const hook of beforeEach) {
+    await _callHook(hook, testContext);
+  }
+
+  await _callTest(test, testContext);
+
+  for (const hook of afterEach) {
+    await _callHook(hook, testContext);
+  }
+};
+
+const _callHook = (hook: Hook, testContext?: TestContext): Promise<any> => {
```

Comment:
it's optional because `beforeAll` and `afterAll` are shared between multiple tests and can't have a context object. Also we're removing it anyway :)